### PR TITLE
AB-3716 Sync banks & clearing numbers with Bankgirot

### PR DESCRIPTION
**Pull request by alexteg-advisa**
This aims to synchronize the clearing number and account number validation with the latest version (`2019-10-22`) of the ["Bank Account Numbers in Swedish Banks" manual](https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf) from Bankgirot.

It **adds** `Klarna Bank` & `MedMera Bank`, **removes** the `Royal Bank of Scotland` and **updates** the clearing number ranges for `Riksgälden` and `Swedbank`. It also **renames** `Amfa Bank` to `Svea Bank` and improves the spelling of some other bank names.

This also adds CI through [Github Actions](https://github.com/features/actions), since that was quick and simple to do and will help ensure all tests pass for PRs. As part of this the `master` branch has been protected to require a PR, review, successful tests & commit message linting before merges to `master`.

AB-3716

### CODE REVIEW

**Comment by alexteg-advisa** [comment]
I pushed some fixups that update the documentation with the added, removed & renamed banks.

File: kontonummer.js [hunk 1]
```diff
@@ -158,6 +158,15 @@
...
             control  : 11
         }
+    },{
+        name    : 'Klarna Bank',
+        regex   : /^(978[0-9])/,
+        modulo  : 11,
+        lengths : {
+            clearing : 4,
+            account  : 7,
+            control  : 11
```
Comment (ghost) [inline kontonummer.js]: If i have understood this correctly, this should be the value of `control`.
`clearing` + `account` - `checkDigit` = `control`.
But this doesn't add up:
`4+7-1=10`

Would it make sense to calculate this value?

**Review by ghost — CHANGES_REQUESTED** [review]
Cool to see GitHub actions in action :D

File: kontonummer.js [hunk 2]
```diff
@@ -33,7 +24,7 @@
             control  : 11
         }
     },{
-        name    : 'Amfa Bank',
+        name    : 'Svea Bank',
         regex   : /^(966[0-9])/,
         modulo  : 11,
         lengths : {
```
Comment (simonbergstrom) [inline kontonummer.js]: Shouldn't control be 11 for this bank according to the new docs? 

Comment 2: "Checksum calculation is made on the entire Clearing number, and seven digits of the actual account number." --> 4 + 7 = 11

File: kontonummer.js [hunk 3]
```diff
@@ -267,13 +276,13 @@
             control  : 10
         }
     },{
-        name    : 'Royal Bank of Scotland',
-        regex   : /^(909[0-9])/,
+        name    : 'Riksgälden',
+        regex   : /^(988[0-9])/,
```
Comment (simonbergstrom) [inline kontonummer.js:200]: For this clearing number I think the control number should be 11 according to the docs. Riksgälden exist for two different interval of clearing numbers.

File: kontonummer.js [hunk 4]
```diff
@@ -203,6 +203,15 @@
...
             control  : 10
         }
+    },{
+        name    : 'MedMera Bank',
+        regex   : /^(965[0-9])/,
+        modulo  : 11,
+        lengths : {
+            clearing : 4,
+            account  : 7,
+            control  : 10
```
Comment (simonbergstrom) [inline kontonummer.js]: Same here with control number. Comment 2 with full clearing + account number as control

File: kontonummer.js [hunk 5]
```diff
@@ -203,6 +203,15 @@
             account  : 7,
             control  : 10
         }
```
Comment (simonbergstrom) [inline kontonummer.js]: Länsförsäkringar for clearing: 9060-9069 is missing in our config. Should it be added?

**Comment by simonbergstrom** [comment]
Went through all banks in our configuration and compared it to the doc and came up with some comments. We can maybe check them together at some point to see if I have understand it correctly.

File: kontonummer.js (hunk 5 above)
Comment (alexteg-advisa) [reply kontonummer.js]: Isn't Länsförsäkringar `9060-9069` covered by `906[0-9]` part of the regex in:

https://github.com/Advisa/kontonummer.js/blob/3fb909fff52c414deec744b59f595cef315b9e77/kontonummer.js#L180-L181

?

File: kontonummer.js (hunk 1 above)
Comment (alexteg-advisa) [reply kontonummer.js]: Sorry for misleading you, but `control` actually refers to the `x` last digits of the _combined_ clearing & account number that should be part of the mod 10/11 checksum calculation, respectively:

### Type 1 account numbers
Comment number:
1. "Clearing number with the exception of the first digit, and seven digits of the actual account number." (basically means the last 10 digits)
2. "Checksum calculation is made on the entire Clearing number, and seven digits of the actual account number." (basically means all the 11 digits)

### Type 2 account numbers
Comment number:
1. "Checksum calculation uses the last ten [10] digits using the modulus 10 check"
2. "Checksum calculation uses the 9 last digits using the modulus 11 check."
3. "Checksum calculation uses the last ten [10] digits using the modulus 10 check, according format for account number (clearing number not included)"

as defined in https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf

File: kontonummer.js (hunk 2 above)
Comment (alexteg-advisa) [reply kontonummer.js]: You are correct. I missed that this was more than just a renaming of a bank.

File: kontonummer.js (hunk 4 above)
Comment (alexteg-advisa) [reply kontonummer.js]: You are totally correct. I got confused by what `control` means in this context.

File: kontonummer.js (hunk 3 above)
Comment (alexteg-advisa) [reply kontonummer.js:200]: You are correct. I got confused by what `control` means in this context.

File: kontonummer.js (hunk 4 above)
Comment (alexteg-advisa) [reply kontonummer.js]: I've fixed this in the fixup commit 2bb6cd2b5d745511c2469864d4e5f03634ef0971

File: kontonummer.js (hunk 3 above)
Comment (alexteg-advisa) [reply kontonummer.js:200]: I've fixed this in the fixup commit feeaf554d2f156f6c1a3bacf43f44653ec7ba04b

File: kontonummer.js (hunk 2 above)
Comment (alexteg-advisa) [reply kontonummer.js]: I've fixed this in the fixup commit c9469c8a966b939e59f89a7288bdd92a0de6b614

**Comment by alexteg-advisa** [comment]
@simonbergstrom @johanadvisa Since both I & Johan got confused regarding what `control` means, I've simplified the length definitions by grouping the 2 different account number types types with a total of 5 variations expressed as comments in the [Bankgirot documentation](https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf) as a constant that just references the type & comment, rather than repeating the same length object.

This makes the code more DRY, makes it more difficult to makes mistakes and maps more clearly to the [Bankgirot documentation](https://www.bankgirot.se/globalassets/dokument/anvandarmanualer/bankernaskontonummeruppbyggnad_anvandarmanual_sv.pdf). See commit: 45177c970c6ce83cafc08c32e97a23ad3c36ada8 `Sync banks & clearing numbers with Bankgirot`. This is similar to how the original repo we forked cleaned up the repeated definitions: https://github.com/jop-io/kontonummer.js/commit/e7abc137343498d2b1264a52c4830ebedd59400c

**Comment by alexteg-advisa** [comment]
Github actions CI running of tests has been split out to a separate PR #5 to enable #4 to be merged without this.

File: kontonummer.js (hunk 5 above)
Comment (simonbergstrom) [reply kontonummer.js]: Aah sorry you are totally correct. 👍

**Review by simonbergstrom — APPROVED** [review]

**Review by ghost — APPROVED** [review]
Adding the constant file makes everything much clearer! Clever!



If the commit order in Github is correct you might need to move the "Group account number formats as in Bankgirot docs" commit before "Add new Klarna & MedMera banks" before squashing.

**Comment by alexteg-advisa** [comment]
The commit "Group account number formats as in Bankgirot docs" (A) is already before the commit "Add new Klarna & MedMera banks" (B), it's just that I created the grouping commit A after B, but moved commit A before B in an interactive rebase and [Github orders commit by author date instead of topologically](https://github.com/isaacs/github/issues/386).

Output of `git log --oneline --graph --no-decorate` (which orders topologically) after squashing the fixup commits:
Screenshot 2020-01-09 at 12 48 31

**Pull request merged**